<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ SCP</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SCP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.5.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/zhanghao-njmu/SCP">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="AnnotateFeatures.html">AnnotateFeatures()</a></code> 
        </dt>
        <dd>AnnotateFeatures</dd>
      </dl><dl><dt>
          
          <code><a href="BBKNN_integrate.html">BBKNN_integrate()</a></code> 
        </dt>
        <dd>BBKNN_integrate</dd>
      </dl><dl><dt>
          
          <code><a href="CC_GenePrefetch.html">CC_GenePrefetch()</a></code> 
        </dt>
        <dd>Prefetch cycle gene</dd>
      </dl><dl><dt>
          
          <code><a href="CSS_integrate.html">CSS_integrate()</a></code> 
        </dt>
        <dd>CSS_integrate</dd>
      </dl><dl><dt>
          
          <code><a href="CellCorHeatmap.html">CellCorHeatmap()</a></code> 
        </dt>
        <dd>CellCorHeatmap</dd>
      </dl><dl><dt>
          
          <code><a href="CellDensityPlot.html">CellDensityPlot()</a></code> 
        </dt>
        <dd>CellDensityPlot</dd>
      </dl><dl><dt>
          
          <code><a href="CellDimPlot.html">CellDimPlot()</a></code> 
        </dt>
        <dd>Visualize cell groups on a 2-dimensional reduction plot</dd>
      </dl><dl><dt>
          
          <code><a href="CellDimPlot3D.html">CellDimPlot3D()</a></code> 
        </dt>
        <dd>3D-Dimensional reduction plot for cell classification visualization.</dd>
      </dl><dl><dt>
          
          <code><a href="CellScoring.html">CellScoring()</a></code> 
        </dt>
        <dd>CellScoring</dd>
      </dl><dl><dt>
          
          <code><a href="CellStatPlot.html">CellStatPlot()</a></code> 
        </dt>
        <dd>Statistical plot of cells</dd>
      </dl><dl><dt>
          
          <code><a href="ComBat_integrate.html">ComBat_integrate()</a></code> 
        </dt>
        <dd>Combat_integrate</dd>
      </dl><dl><dt>
          
          <code><a href="Conos_integrate.html">Conos_integrate()</a></code> 
        </dt>
        <dd>Conos_integrate</dd>
      </dl><dl><dt>
          
          <code><a href="CreateDataFile.html">CreateDataFile()</a></code> 
        </dt>
        <dd>CreateDataFile</dd>
      </dl><dl><dt>
          
          <code><a href="CreateMetaFile.html">CreateMetaFile()</a></code> 
        </dt>
        <dd>CreateMetaFile</dd>
      </dl><dl><dt>
          
          <code><a href="DefaultReduction.html">DefaultReduction()</a></code> 
        </dt>
        <dd>Find the default reduction name in a Seurat object.</dd>
      </dl><dl><dt>
          
          <code><a href="DynamicHeatmap.html">DynamicHeatmap()</a></code> 
        </dt>
        <dd>Heatmap plot for dynamic features along lineages</dd>
      </dl><dl><dt>
          
          <code><a href="DynamicPlot.html">DynamicPlot()</a></code> 
        </dt>
        <dd>Plot for dynamic features along lineages.</dd>
      </dl><dl><dt>
          
          <code><a href="EnrichmentPlot.html">EnrichmentPlot()</a></code> 
        </dt>
        <dd>EnrichmentPlot</dd>
      </dl><dl><dt>
          
          <code><a href="FeatureCorPlot.html">FeatureCorPlot()</a></code> 
        </dt>
        <dd>Features correlation plot</dd>
      </dl><dl><dt>
          
          <code><a href="FeatureDimPlot.html">FeatureDimPlot()</a></code> 
        </dt>
        <dd>Visualize feature values on a 2-dimensional reduction plot</dd>
      </dl><dl><dt>
          
          <code><a href="FeatureDimPlot3D.html">FeatureDimPlot3D()</a></code> 
        </dt>
        <dd>3D-Dimensional reduction plot for gene expression visualization.</dd>
      </dl><dl><dt>
          
          <code><a href="FeatureHeatmap.html">FeatureHeatmap()</a></code> 
        </dt>
        <dd>FeatureHeatmap</dd>
      </dl><dl><dt>
          
          <code><a href="FeatureStatPlot.html">FeatureStatPlot()</a></code> 
        </dt>
        <dd>Statistical plot of features</dd>
      </dl><dl><dt>
          
          <code><a href="FetchH5.html">FetchH5()</a></code> 
        </dt>
        <dd>Fetch data from the hdf5 file</dd>
      </dl><dl><dt>
          
          <code><a href="GSEAPlot.html">GSEAPlot()</a></code> 
        </dt>
        <dd>GSEA Plot</dd>
      </dl><dl><dt>
          
          <code><a href="GeneConvert.html">GeneConvert()</a></code> 
        </dt>
        <dd>Gene ID conversion function using biomart</dd>
      </dl><dl><dt>
          
          <code><a href="GraphPlot.html">GraphPlot()</a></code> 
        </dt>
        <dd>GraphPlot</dd>
      </dl><dl><dt>
          
          <code><a href="GroupHeatmap.html">GroupHeatmap()</a></code> 
        </dt>
        <dd>GroupHeatmap</dd>
      </dl><dl><dt>
          
          <code><a href="Harmony_integrate.html">Harmony_integrate()</a></code> 
        </dt>
        <dd>Harmony_integrate</dd>
      </dl><dl><dt>
          
          <code><a href="Integration_SCP.html">Integration_SCP()</a></code> 
        </dt>
        <dd>Integration_SCP</dd>
      </dl><dl><dt>
          
          <code><a href="LIGER_integrate.html">LIGER_integrate()</a></code> 
        </dt>
        <dd>LIGER_integrate</dd>
      </dl><dl><dt>
          
          <code><a href="LineagePlot.html">LineagePlot()</a></code> 
        </dt>
        <dd>LineagePlot</dd>
      </dl><dl><dt>
          
          <code><a href="ListDB.html">ListDB()</a></code> 
        </dt>
        <dd>ListDB</dd>
      </dl><dl><dt>
          
          <code><a href="MNN_integrate.html">MNN_integrate()</a></code> 
        </dt>
        <dd>MNN_integrate</dd>
      </dl><dl><dt>
          
          <code><a href="PAGAPlot.html">PAGAPlot()</a></code> 
        </dt>
        <dd>PAGA plot</dd>
      </dl><dl><dt>
          
          <code><a href="PrepareDB.html">PrepareDB()</a></code> 
        </dt>
        <dd>Prepare the gene annotation databases</dd>
      </dl><dl><dt>
          
          <code><a href="PrepareEnv.html">PrepareEnv()</a></code> 
        </dt>
        <dd>Prepare SCP python environment</dd>
      </dl><dl><dt>
          
          <code><a href="PrepareSCExplorer.html">PrepareSCExplorer()</a></code> 
        </dt>
        <dd>Prepare Seurat object data for the SCExplorer into hdf5 file</dd>
      </dl><dl><dt>
          
          <code><a href="ProjectionPlot.html">ProjectionPlot()</a></code> 
        </dt>
        <dd>Projection Plot</dd>
      </dl><dl><dt>
          
          <code><a href="RecoverCounts.html">RecoverCounts()</a></code> 
        </dt>
        <dd>Attempt to recover raw counts from the normalized matrix.</dd>
      </dl><dl><dt>
          
          <code><a href="RenameClusters.html">RenameClusters()</a></code> 
        </dt>
        <dd>Rename clusters for the Seurat object</dd>
      </dl><dl><dt>
          
          <code><a href="RenameFeatures.html">RenameFeatures()</a></code> 
        </dt>
        <dd>Rename features for the Seurat object</dd>
      </dl><dl><dt>
          
          <code><a href="RunCSSMap.html">RunCSSMap()</a></code> 
        </dt>
        <dd>Single-cell reference mapping with CSS method</dd>
      </dl><dl><dt>
          
          <code><a href="RunCellQC.html">RunCellQC()</a></code> 
        </dt>
        <dd>Run cell-level quality control for single cell RNA-seq data.</dd>
      </dl><dl><dt>
          
          <code><a href="RunDEtest.html">RunDEtest()</a></code> 
        </dt>
        <dd>Differential gene test</dd>
      </dl><dl><dt>
          
          <code><a href="RunDM.html">RunDM()</a></code> 
        </dt>
        <dd>Run DiffusionMap(DM)</dd>
      </dl><dl><dt>
          
          <code><a href="RunDimReduction.html">RunDimReduction()</a></code> 
        </dt>
        <dd>Run dimensionality reduction</dd>
      </dl><dl><dt>
          
          <code><a href="RunDoubletCalling.html">RunDoubletCalling()</a></code> 
        </dt>
        <dd>Run doublet-calling for single cell RNA-seq data.</dd>
      </dl><dl><dt>
          
          <code><a href="RunDynamicEnrichment.html">RunDynamicEnrichment()</a></code> 
        </dt>
        <dd>RunDynamicEnrichment</dd>
      </dl><dl><dt>
          
          <code><a href="RunDynamicFeatures.html">RunDynamicFeatures()</a></code> 
        </dt>
        <dd>RunDynamicFeatures</dd>
      </dl><dl><dt>
          
          <code><a href="RunEnrichment.html">RunEnrichment()</a></code> 
        </dt>
        <dd>Perform the enrichment analysis(over-representation) on the genes</dd>
      </dl><dl><dt>
          
          <code><a href="RunGLMPCA.html">RunGLMPCA()</a></code> 
        </dt>
        <dd>Run GLMPCA</dd>
      </dl><dl><dt>
          
          <code><a href="RunGSEA.html">RunGSEA()</a></code> 
        </dt>
        <dd>Perform the enrichment analysis(GSEA) on the genes</dd>
      </dl><dl><dt>
          
          <code><a href="RunHarmony2.html">RunHarmony2()</a></code> 
        </dt>
        <dd>Harmony single cell integration</dd>
      </dl><dl><dt>
          
          <code><a href="RunKNNMap.html">RunKNNMap()</a></code> 
        </dt>
        <dd>Single-cell reference mapping with KNN method</dd>
      </dl><dl><dt>
          
          <code><a href="RunKNNPredict.html">RunKNNPredict()</a></code> 
        </dt>
        <dd>Annotate single cells using similarity.</dd>
      </dl><dl><dt>
          
          <code><a href="RunLargeVis.html">RunLargeVis()</a></code> 
        </dt>
        <dd>Run LargeVis</dd>
      </dl><dl><dt>
          
          <code><a href="RunMDS.html">RunMDS()</a></code> 
        </dt>
        <dd>Run multi-dimensional scaling(MDS)</dd>
      </dl><dl><dt>
          
          <code><a href="RunMonocle2.html">RunMonocle2()</a></code> 
        </dt>
        <dd>Run Monocle2 analysis</dd>
      </dl><dl><dt>
          
          <code><a href="RunMonocle3.html">RunMonocle3()</a></code> 
        </dt>
        <dd>Run Monocle3 analysis</dd>
      </dl><dl><dt>
          
          <code><a href="RunNMF.html">RunNMF()</a></code> 
        </dt>
        <dd>Run Non-negative matrix factorization(NMF)</dd>
      </dl><dl><dt>
          
          <code><a href="RunPAGA.html">RunPAGA()</a></code> 
        </dt>
        <dd>Run PAGA analysis</dd>
      </dl><dl><dt>
          
          <code><a href="RunPCAMap.html">RunPCAMap()</a></code> 
        </dt>
        <dd>Single-cell reference mapping with PCA method</dd>
      </dl><dl><dt>
          
          <code><a href="RunPHATE.html">RunPHATE()</a></code> 
        </dt>
        <dd>Run PHATE</dd>
      </dl><dl><dt>
          
          <code><a href="RunPaCMAP.html">RunPaCMAP()</a></code> 
        </dt>
        <dd>Run PaCMAP</dd>
      </dl><dl><dt>
          
          <code><a href="RunPalantir.html">RunPalantir()</a></code> 
        </dt>
        <dd>Run Palantir analysis</dd>
      </dl><dl><dt>
          
          <code><a href="RunSCExplorer.html">RunSCExplorer()</a></code> 
        </dt>
        <dd>RunSCExplorer</dd>
      </dl><dl><dt>
          
          <code><a href="RunSCVELO.html">RunSCVELO()</a></code> 
        </dt>
        <dd>Run scVelo analysis</dd>
      </dl><dl><dt>
          
          <code><a href="RunScmap.html">RunScmap()</a></code> 
        </dt>
        <dd>Annotate single cells using scmap.</dd>
      </dl><dl><dt>
          
          <code><a href="RunSeuratMap.html">RunSeuratMap()</a></code> 
        </dt>
        <dd>Single-cell reference mapping with Seurat method</dd>
      </dl><dl><dt>
          
          <code><a href="RunSingleR.html">RunSingleR()</a></code> 
        </dt>
        <dd>Annotate single cells using scmap.</dd>
      </dl><dl><dt>
          
          <code><a href="RunSlingshot.html">RunSlingshot()</a></code> 
        </dt>
        <dd>RunSlingshot</dd>
      </dl><dl><dt>
          
          <code><a href="RunSymphonyMap.html">RunSymphonyMap()</a></code> 
        </dt>
        <dd>Single-cell reference mapping with Symphony method</dd>
      </dl><dl><dt>
          
          <code><a href="RunTriMap.html">RunTriMap()</a></code> 
        </dt>
        <dd>Run TriMap</dd>
      </dl><dl><dt>
          
          <code><a href="RunUMAP2.html">RunUMAP2()</a></code> 
        </dt>
        <dd>Run UMAP</dd>
      </dl><dl><dt>
          
          <code><a href="SankeyPlot.html">SankeyPlot()</a></code> 
        </dt>
        <dd>SankeyPlot</dd>
      </dl><dl><dt>
          
          <code><a href="Scanorama_integrate.html">Scanorama_integrate()</a></code> 
        </dt>
        <dd>Scanorama_integrate</dd>
      </dl><dl><dt>
          
          <code><a href="Seurat_integrate.html">Seurat_integrate()</a></code> 
        </dt>
        <dd>Seurat_integrate</dd>
      </dl><dl><dt>
          
          <code><a href="SrtAppend.html">SrtAppend()</a></code> 
        </dt>
        <dd>Append a Seurat object to another</dd>
      </dl><dl><dt>
          
          <code><a href="SrtReorder.html">SrtReorder()</a></code> 
        </dt>
        <dd>Reorder idents by the gene expression</dd>
      </dl><dl><dt>
          
          <code><a href="Standard_SCP.html">Standard_SCP()</a></code> 
        </dt>
        <dd>Standard_SCP</dd>
      </dl><dl><dt>
          
          <code><a href="StatPlot.html">StatPlot()</a></code> 
        </dt>
        <dd>StatPlot</dd>
      </dl><dl><dt>
          
          <code><a href="Uncorrected_integrate.html">Uncorrected_integrate()</a></code> 
        </dt>
        <dd>Uncorrected_integrate</dd>
      </dl><dl><dt>
          
          <code><a href="VelocityPlot.html">VelocityPlot()</a></code> 
        </dt>
        <dd>Velocity Plot</dd>
      </dl><dl><dt>
          
          <code><a href="VolcanoPlot.html">VolcanoPlot()</a></code> 
        </dt>
        <dd>VolcanoPlot</dd>
      </dl><dl><dt>
          
          <code><a href="adata_to_srt.html">adata_to_srt()</a></code> 
        </dt>
        <dd>Convert an anndata object to a seurat object using reticulate</dd>
      </dl><dl><dt>
          
          <code><a href="adjcolors.html">adjcolors()</a></code> 
        </dt>
        <dd>Convert a color with arbitrary transparency to a fixed color</dd>
      </dl><dl><dt>
          
          <code><a href="as_matrix.html">as_matrix()</a></code> 
        </dt>
        <dd>Attempts to turn a dgCMatrix into a dense matrix</dd>
      </dl><dl><dt>
          
          <code><a href="blendcolors.html">blendcolors()</a></code> 
        </dt>
        <dd>Blend colors</dd>
      </dl><dl><dt>
          
          <code><a href="capitalize.html">capitalize()</a></code> 
        </dt>
        <dd>Capitalizes the characters
Making the first letter uppercase</dd>
      </dl><dl><dt>
          
          <code><a href="check_DataType.html">check_DataType()</a></code> 
        </dt>
        <dd>Check and report the type of data</dd>
      </dl><dl><dt>
          
          <code><a href="check_Python.html">check_Python()</a></code> 
        </dt>
        <dd>Check and install python packages</dd>
      </dl><dl><dt>
          
          <code><a href="check_R.html">check_R()</a></code> 
        </dt>
        <dd>Check and install R packages</dd>
      </dl><dl><dt>
          
          <code><a href="check_srtList.html">check_srtList()</a></code> 
        </dt>
        <dd>Check and preprocess a list of seurat objects</dd>
      </dl><dl><dt>
          
          <code><a href="check_srtMerge.html">check_srtMerge()</a></code> 
        </dt>
        <dd>Check and preprocess a merged seurat object</dd>
      </dl><dl><dt>
          
          <code><a href="compute_velocity_on_grid.html">compute_velocity_on_grid()</a></code> 
        </dt>
        <dd>Compute velocity on grid
The original python code is on https://github.com/theislab/scvelo/blob/master/scvelo/plotting/velocity_embedding_grid.py</dd>
      </dl><dl><dt>
          
          <code><a href="conda_install.html">conda_install()</a></code> 
        </dt>
        <dd>Installs a list of packages into a specified conda environment</dd>
      </dl><dl><dt>
          
          <code><a href="conda_python.html">conda_python()</a></code> 
        </dt>
        <dd>Find the path to Python associated with a conda environment</dd>
      </dl><dl><dt>
          
          <code><a href="db_DoubletDetection.html">db_DoubletDetection()</a></code> 
        </dt>
        <dd>Run doublet-calling with DoubletDetection</dd>
      </dl><dl><dt>
          
          <code><a href="db_Scrublet.html">db_Scrublet()</a></code> 
        </dt>
        <dd>Run doublet-calling with Scrublet</dd>
      </dl><dl><dt>
          
          <code><a href="db_scDblFinder.html">db_scDblFinder()</a></code> 
        </dt>
        <dd>Run doublet-calling with scDblFinder</dd>
      </dl><dl><dt>
          
          <code><a href="db_scds.html">db_scds()</a></code> 
        </dt>
        <dd>Run doublet-calling with scds</dd>
      </dl><dl><dt>
          
          <code><a href="download.html">download()</a></code> 
        </dt>
        <dd>Download File from the Internet</dd>
      </dl><dl><dt>
          
          <code><a href="drop_data.html">drop_data()</a></code> 
        </dt>
        <dd>Drop all data in the plot (only one observation is kept)</dd>
      </dl><dl><dt>
          
          <code><a href="env_exist.html">env_exist()</a></code> 
        </dt>
        <dd>Check if a conda environment exists</dd>
      </dl><dl><dt>
          
          <code><a href="exist_Python_pkgs.html">exist_Python_pkgs()</a></code> 
        </dt>
        <dd>Check if the python package exists in the environment</dd>
      </dl><dl><dt>
          
          <code><a href="fastMNN_integrate.html">fastMNN_integrate()</a></code> 
        </dt>
        <dd>fastMNN_integrate</dd>
      </dl><dl><dt>
          
          <code><a href="find_conda.html">find_conda()</a></code> 
        </dt>
        <dd>Find an appropriate conda binary</dd>
      </dl><dl><dt>
          
          <code><a href="geom_alluvial.html">geom_alluvial()</a></code> 
        </dt>
        <dd>geom_alluvial</dd>
      </dl><dl><dt>
          
          <code><a href="geom_alluvial_label.html">geom_alluvial_text()</a></code> <code><a href="geom_alluvial_label.html">geom_alluvial_label()</a></code> 
        </dt>
        <dd>geom_alluvial_label</dd>
      </dl><dl><dt>
          
          <code><a href="geom_sankey.html">geom_sankey()</a></code> 
        </dt>
        <dd>geom_sankey

Creates a sankey plot which visualize flows between nodes. Each observation needs to have a `x` aesthetic as well as a `next_x` column which declares where that observation should flow.
Also each observation should have a `node` and a `next_node` aesthetic which provide information about which group in the y-direction. By default each row of the data frame is counted to calculate the size of flows. A manual flow value can be added with the `value` aesthetic.</dd>
      </dl><dl><dt>
          
          <code><a href="geom_sankey_bump.html">geom_sankey_bump()</a></code> 
        </dt>
        <dd>geom_sankey_bump</dd>
      </dl><dl><dt>
          
          <code><a href="geom_sankey_label.html">geom_sankey_label()</a></code> <code><a href="geom_sankey_label.html">geom_sankey_text()</a></code> 
        </dt>
        <dd>geom_sankey_label</dd>
      </dl><dl><dt>
          
          <code><a href="get_vars.html">get_vars()</a></code> 
        </dt>
        <dd>Get used vars in a ggplot object</dd>
      </dl><dl><dt>
          
          <code><a href="ifnb_sub.html">ifnb_sub</a></code> 
        </dt>
        <dd>A subsetted version of 'ifnb' datasets</dd>
      </dl><dl><dt>
          
          <code><a href="invoke.html">invoke()</a></code> 
        </dt>
        <dd>Invoke a function with a list of arguments</dd>
      </dl><dl><dt>
          
          <code><a href="isOutlier.html">isOutlier()</a></code> 
        </dt>
        <dd>Detect outliers using MAD(Median and Median Absolute Deviation) method</dd>
      </dl><dl><dt>
          
          <code><a href="iterchunks.html">iterchunks()</a></code> 
        </dt>
        <dd>Generate a iterator along chunks of a vector</dd>
      </dl><dl><dt>
          
          <code><a href="make_long.html">make_long()</a></code> 
        </dt>
        <dd>make_long</dd>
      </dl><dl><dt>
          
          <code><a href="palette_scp.html">palette_scp()</a></code> 
        </dt>
        <dd>Color palettes collected in SCP.</dd>
      </dl><dl><dt>
          
          <code><a href="panc8_sub.html">panc8_sub</a></code> 
        </dt>
        <dd>A subsetted version of human 'panc8' datasets</dd>
      </dl><dl><dt>
          
          <code><a href="pancreas_sub.html">pancreas_sub</a></code> 
        </dt>
        <dd>A subsetted version of mouse 'pancreas' datasets</dd>
      </dl><dl><dt>
          
          <code><a href="panel_fix.html">panel_fix()</a></code> 
        </dt>
        <dd>Set the panel width/height of a plot object to a fixed value.</dd>
      </dl><dl><dt>
          
          <code><a href="panel_fix_single.html">panel_fix_single()</a></code> 
        </dt>
        <dd>Fix the width/height of panels in a single plot object.</dd>
      </dl><dl><dt>
          
          <code><a href="scVI_integrate.html">scVI_integrate()</a></code> 
        </dt>
        <dd>scVI_integrate</dd>
      </dl><dl><dt>
          
          <code><a href="segementsDf.html">segementsDf()</a></code> 
        </dt>
        <dd>Shorten and offset the segment</dd>
      </dl><dl><dt>
          
          <code><a href="show_palettes.html">show_palettes()</a></code> 
        </dt>
        <dd>Show the color palettes</dd>
      </dl><dl><dt>
          
          <code><a href="slim_data.html">slim_data()</a></code> 
        </dt>
        <dd>Drop unused data from the plot to reduce the object size</dd>
      </dl><dl><dt>
          
          <code><a href="srt_to_adata.html">srt_to_adata()</a></code> 
        </dt>
        <dd>Convert a seurat object to an anndata object using reticulate</dd>
      </dl><dl><dt>
          
          <code><a href="theme_blank.html">theme_blank()</a></code> 
        </dt>
        <dd>Blank theme</dd>
      </dl><dl><dt>
          
          <code><a href="theme_sankey.html">theme_sankey()</a></code> <code><a href="theme_sankey.html">theme_alluvial()</a></code> <code><a href="theme_sankey.html">theme_sankey_bump()</a></code> 
        </dt>
        <dd>sankey_themes</dd>
      </dl><dl><dt>
          
          <code><a href="theme_scp.html">theme_scp()</a></code> 
        </dt>
        <dd>SCP theme</dd>
      </dl><dl><dt>
          
          <code><a href="tochunks.html">tochunks()</a></code> 
        </dt>
        <dd>Split a vector into the chunks</dd>
      </dl><dl><dt>
          
          <code><a href="unnest.html">unnest()</a></code> 
        </dt>
        <dd>Implement similar functions to the <code>unnest</code> function in the tidyr package</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Hao Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

