<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function performs KNN prediction to annotate cell types based on reference scRNA-seq or bulk RNA-seq data."><title>RunKNNPredict — RunKNNPredict • SCP</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="RunKNNPredict — RunKNNPredict"><meta property="og:description" content="This function performs KNN prediction to annotate cell types based on reference scRNA-seq or bulk RNA-seq data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SCP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/zhanghao-njmu/SCP">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>RunKNNPredict</h1>
      <small class="dont-index">Source: <a href="https://github.com/zhanghao-njmu/SCP/blob/HEAD/R/SCP-cell_annotation.R" class="external-link"><code>R/SCP-cell_annotation.R</code></a></small>
      <div class="d-none name"><code>RunKNNPredict.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function performs KNN prediction to annotate cell types based on reference scRNA-seq or bulk RNA-seq data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">RunKNNPredict</span><span class="op">(</span></span>
<span>  <span class="va">srt_query</span>,</span>
<span>  srt_ref <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bulk_ref <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  query_group <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ref_group <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  query_assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ref_assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  query_reduction <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ref_reduction <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  query_dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>  ref_dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>  query_collapsing <span class="op">=</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">query_group</span><span class="op">)</span>,</span>
<span>  ref_collapsing <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  return_full_distance_matrix <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  features_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HVF"</span>, <span class="st">"DE"</span><span class="op">)</span>,</span>
<span>  feature_source <span class="op">=</span> <span class="st">"both"</span>,</span>
<span>  nfeatures <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  DEtest_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>max.cells.per.ident <span class="op">=</span> <span class="fl">200</span>, test.use <span class="op">=</span> <span class="st">"wilcox"</span><span class="op">)</span>,</span>
<span>  DE_threshold <span class="op">=</span> <span class="st">"p_val_adj &lt; 0.05"</span>,</span>
<span>  nn_method <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  distance_metric <span class="op">=</span> <span class="st">"cosine"</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  filter_lowfreq <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">"KNNPredict"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>srt_query</dt>
<dd><p>An object of class Seurat to be annotated with cell types.</p></dd>


<dt>srt_ref</dt>
<dd><p>An object of class Seurat storing the reference cells.</p></dd>


<dt>bulk_ref</dt>
<dd><p>A cell atlas matrix, where cell types are represented by columns and genes are represented by rows, for example, SCP::ref_scHCL. Either `srt_ref` or `bulk_ref` must be provided.</p></dd>


<dt>query_group</dt>
<dd><p>A character vector specifying the column name in the `srt_query` metadata that represents the cell grouping.</p></dd>


<dt>ref_group</dt>
<dd><p>A character vector specifying the column name in the `srt_ref` metadata that represents the cell grouping.</p></dd>


<dt>query_assay</dt>
<dd><p>A character vector specifying the assay to be used for the query data. Defaults to the default assay of the `srt_query` object.</p></dd>


<dt>ref_assay</dt>
<dd><p>A character vector specifying the assay to be used for the reference data. Defaults to the default assay of the `srt_ref` object.</p></dd>


<dt>query_reduction</dt>
<dd><p>A character vector specifying the dimensionality reduction method used for the query data. If NULL, the function will use the default reduction method specified in the `srt_query` object.</p></dd>


<dt>ref_reduction</dt>
<dd><p>A character vector specifying the dimensionality reduction method used for the reference data. If NULL, the function will use the default reduction method specified in the `srt_ref` object.</p></dd>


<dt>query_dims</dt>
<dd><p>A numeric vector specifying the dimensions to be used for the query data. Defaults to the first 30 dimensions.</p></dd>


<dt>ref_dims</dt>
<dd><p>A numeric vector specifying the dimensions to be used for the reference data. Defaults to the first 30 dimensions.</p></dd>


<dt>query_collapsing</dt>
<dd><p>A boolean value indicating whether the query data should be collapsed to group-level average expression values. If TRUE, the function will calculate the average expression values for each group in the query data and the annotation will be performed separately for each group. Otherwise it will use the raw expression values for each cell.</p></dd>


<dt>ref_collapsing</dt>
<dd><p>A boolean value indicating whether the reference data should be collapsed to group-level average expression values. If TRUE, the function will calculate the average expression values for each group in the reference data and the annotation will be performed separately for each group. Otherwise it will use the raw expression values for each cell.</p></dd>


<dt>return_full_distance_matrix</dt>
<dd><p>A boolean value indicating whether the full distance matrix should be returned. If TRUE, the function will return the distance matrix used for the KNN prediction, otherwise it will only return the annotated cell types.</p></dd>


<dt>features</dt>
<dd><p>A character vector specifying the features (genes) to be used for the KNN prediction. If NULL, all the features in the query and reference data will be used.</p></dd>


<dt>features_type</dt>
<dd><p>A character vector specifying the type of features to be used for the KNN prediction. Must be one of "HVF" (highly variable features) or "DE" (differentially expressed features). Defaults to "HVF".</p></dd>


<dt>feature_source</dt>
<dd><p>A character vector specifying the source of the features to be used for the KNN prediction. Must be one of "both", "query", or "ref". Defaults to "both".</p></dd>


<dt>nfeatures</dt>
<dd><p>An integer specifying the maximum number of features to be used for the KNN prediction. Defaults to 2000.</p></dd>


<dt>DEtest_param</dt>
<dd><p>A list of parameters to be passed to the differential expression test function if `features_type` is set to "DE". Defaults to `list(max.cells.per.ident = 200, test.use = "wilcox")`.</p></dd>


<dt>DE_threshold</dt>
<dd><p>Threshold used to filter the DE features. Default is "p_val &lt; 0.05". If using "roc" test, <code>DE_threshold</code> should be needs to be reassigned. e.g. "power &gt; 0.5".</p></dd>


<dt>nn_method</dt>
<dd><p>A character vector specifying the method to be used for finding nearest neighbors. Must be one of "raw", "rann", or "annoy". Defaults to "raw".</p></dd>


<dt>distance_metric</dt>
<dd><p>A character vector specifying the distance metric to be used for calculating similarity between cells. Must be one of "cosine", "euclidean", "manhattan", or "hamming". Defaults to "cosine".</p></dd>


<dt>k</dt>
<dd><p>An integer specifying the number of nearest neighbors to be considered for the KNN prediction. Defaults to 30.</p></dd>


<dt>filter_lowfreq</dt>
<dd><p>An integer specifying the threshold for filtering low-frequency cell types from the predicted results. Cell types with a frequency lower than `filter_lowfreq` will be labelled as "unreliable". Defaults to 0, which means no filtering will be performed.</p></dd>


<dt>prefix</dt>
<dd><p>A character vector specifying the prefix to be added to the resulting annotations. Defaults to "KNNPredict".</p></dd>

</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Annotate cells using bulk RNA-seq data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pancreas_sub"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ref_scMCA"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="Standard_SCP.html">Standard_SCP</a></span><span class="op">(</span><span class="va">pancreas_sub</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:09.372655] Start Standard_SCP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:09.374415] Checking srtList... ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Data 1/1 of the srtList is raw_counts. Perform NormalizeData(LogNormalize) on the data ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Perform FindVariableFeatures on the data 1/1 of the srtList...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Use the separate HVF from srtList...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of available HVF: 2000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:10.354659] Finished checking.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:10.354907] Perform ScaleData on the data...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:10.472768] Perform linear dimension reduction (pca) on the data...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:11.186693] Perform FindClusters (louvain) on the data...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:11.2534] Reorder clusters...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:11.341639] Perform nonlinear dimension reduction (umap) on the data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Non-linear dimensionality reduction(umap) using Reduction(Standardpca, dims:1-13) as input</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘spam’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘spam’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Non-linear dimensionality reduction(umap) using Reduction(Standardpca, dims:1-13) as input</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘spam’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘spam’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:22.519953] Standard_SCP done</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Elapsed time: 13.15 secs </span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu">RunKNNPredict</span><span class="op">(</span>srt_query <span class="op">=</span> <span class="va">pancreas_sub</span>, bulk_ref <span class="op">=</span> <span class="va">ref_scMCA</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 535 features to calculate distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected query data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected reference data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate similarity...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 'raw' method to find neighbors.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Predict cell type...</span>
<span class="r-in"><span><span class="fu"><a href="CellDimPlot.html">CellDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, group.by <span class="op">=</span> <span class="st">"KNNPredict_classification"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Removal of low credible cell types from the predicted results</span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu">RunKNNPredict</span><span class="op">(</span>srt_query <span class="op">=</span> <span class="va">pancreas_sub</span>, bulk_ref <span class="op">=</span> <span class="va">ref_scMCA</span>, filter_lowfreq <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 535 features to calculate distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected query data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected reference data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate similarity...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 'raw' method to find neighbors.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Predict cell type...</span>
<span class="r-in"><span><span class="fu"><a href="CellDimPlot.html">CellDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, group.by <span class="op">=</span> <span class="st">"KNNPredict_classification"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Annotate clusters using bulk RNA-seq data</span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu">RunKNNPredict</span><span class="op">(</span>srt_query <span class="op">=</span> <span class="va">pancreas_sub</span>, query_group <span class="op">=</span> <span class="st">"SubCellType"</span>, bulk_ref <span class="op">=</span> <span class="va">ref_scMCA</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 535 features to calculate distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected query data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected reference data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate similarity...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 'raw' method to find neighbors.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Predict cell type...</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> No cell overlap between new meta data and Seurat object</span>
<span class="r-in"><span><span class="fu"><a href="CellDimPlot.html">CellDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, group.by <span class="op">=</span> <span class="st">"KNNPredict_classification"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Annotate using single cell RNA-seq data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"panc8_sub"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Simply convert genes from human to mouse and preprocess the data</span></span></span>
<span class="r-in"><span><span class="va">genenames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.unique.html" class="external-link">make.unique</a></span><span class="op">(</span><span class="fu"><a href="capitalize.html">capitalize</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">panc8_sub</span><span class="op">)</span>, force_tolower <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">panc8_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="RenameFeatures.html">RenameFeatures</a></span><span class="op">(</span><span class="va">panc8_sub</span>, newnames <span class="op">=</span> <span class="va">genenames</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Rename features for the assay: RNA</span>
<span class="r-in"><span><span class="va">panc8_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="check_srtMerge.html">check_srtMerge</a></span><span class="op">(</span><span class="va">panc8_sub</span>, batch <span class="op">=</span> <span class="st">"tech"</span><span class="op">)</span><span class="op">[[</span><span class="st">"srtMerge"</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:26.682302] Spliting srtMerge into srtList by column tech... ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:27.08777] Checking srtList... ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Data 1/5 of the srtList is raw_normalized_counts. Perform NormalizeData(LogNormalize) on the data ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Perform FindVariableFeatures on the data 1/5 of the srtList...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Data 2/5 of the srtList is raw_normalized_counts. Perform NormalizeData(LogNormalize) on the data ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Perform FindVariableFeatures on the data 2/5 of the srtList...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Data 3/5 of the srtList is raw_normalized_counts. Perform NormalizeData(LogNormalize) on the data ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Perform FindVariableFeatures on the data 3/5 of the srtList...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Data 4/5 of the srtList is raw_counts. Perform NormalizeData(LogNormalize) on the data ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Perform FindVariableFeatures on the data 4/5 of the srtList...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Data 5/5 of the srtList is raw_counts. Perform NormalizeData(LogNormalize) on the data ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Perform FindVariableFeatures on the data 5/5 of the srtList...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Use the separate HVF from srtList...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of available HVF: 2000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:29.873348] Finished checking.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu">RunKNNPredict</span><span class="op">(</span>srt_query <span class="op">=</span> <span class="va">pancreas_sub</span>, srt_ref <span class="op">=</span> <span class="va">panc8_sub</span>, ref_group <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the HVF to calculate distance metric.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 631 features to calculate distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected query data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected reference data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate similarity...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 'raw' method to find neighbors.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Predict cell type...</span>
<span class="r-in"><span><span class="fu"><a href="CellDimPlot.html">CellDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, group.by <span class="op">=</span> <span class="st">"KNNPredict_classification"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-4.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="FeatureDimPlot.html">FeatureDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, features <span class="op">=</span> <span class="st">"KNNPredict_simil"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu">RunKNNPredict</span><span class="op">(</span></span></span>
<span class="r-in"><span>  srt_query <span class="op">=</span> <span class="va">pancreas_sub</span>, srt_ref <span class="op">=</span> <span class="va">panc8_sub</span>,</span></span>
<span class="r-in"><span>  ref_group <span class="op">=</span> <span class="st">"celltype"</span>, ref_collapsing <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the HVF to calculate distance metric.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 631 features to calculate distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected query data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected reference data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate similarity...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 'raw' method to find neighbors.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Predict cell type...</span>
<span class="r-in"><span><span class="fu"><a href="CellDimPlot.html">CellDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, group.by <span class="op">=</span> <span class="st">"KNNPredict_classification"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-6.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="FeatureDimPlot.html">FeatureDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, features <span class="op">=</span> <span class="st">"KNNPredict_prob"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-7.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu">RunKNNPredict</span><span class="op">(</span></span></span>
<span class="r-in"><span>  srt_query <span class="op">=</span> <span class="va">pancreas_sub</span>, srt_ref <span class="op">=</span> <span class="va">panc8_sub</span>,</span></span>
<span class="r-in"><span>  query_group <span class="op">=</span> <span class="st">"SubCellType"</span>, ref_group <span class="op">=</span> <span class="st">"celltype"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the HVF to calculate distance metric.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 631 features to calculate distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected query data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected reference data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate similarity...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 'raw' method to find neighbors.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Predict cell type...</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> No cell overlap between new meta data and Seurat object</span>
<span class="r-in"><span><span class="fu"><a href="CellDimPlot.html">CellDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, group.by <span class="op">=</span> <span class="st">"KNNPredict_classification"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-8.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="FeatureDimPlot.html">FeatureDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, features <span class="op">=</span> <span class="st">"KNNPredict_simil"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-9.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Annotate with DE gene instead of HVF</span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu">RunKNNPredict</span><span class="op">(</span></span></span>
<span class="r-in"><span>  srt_query <span class="op">=</span> <span class="va">pancreas_sub</span>, srt_ref <span class="op">=</span> <span class="va">panc8_sub</span>,</span></span>
<span class="r-in"><span>  ref_group <span class="op">=</span> <span class="st">"celltype"</span>,</span></span>
<span class="r-in"><span>  features_type <span class="op">=</span> <span class="st">"DE"</span>, feature_source <span class="op">=</span> <span class="st">"ref"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:34.234467] Start DEtest</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Workers: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Find all markers(wilcox) among 13 groups...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                                                                      |                                                                                                              |   0%  |                                                                                                                      |==============================================================================================================| 100%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:44.597592] DEtest done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Elapsed time:10.36 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the DE features from AllMarkers_wilcox to calculate distance metric.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DE features number of the ref data: 2000</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 1865 features to calculate distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected query data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected reference data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate similarity...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 'raw' method to find neighbors.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Predict cell type...</span>
<span class="r-in"><span><span class="fu"><a href="CellDimPlot.html">CellDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, group.by <span class="op">=</span> <span class="st">"KNNPredict_classification"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-10.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="FeatureDimPlot.html">FeatureDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, features <span class="op">=</span> <span class="st">"KNNPredict_simil"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-11.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu">RunKNNPredict</span><span class="op">(</span></span></span>
<span class="r-in"><span>  srt_query <span class="op">=</span> <span class="va">pancreas_sub</span>, srt_ref <span class="op">=</span> <span class="va">panc8_sub</span>,</span></span>
<span class="r-in"><span>  query_group <span class="op">=</span> <span class="st">"SubCellType"</span>, ref_group <span class="op">=</span> <span class="st">"celltype"</span>,</span></span>
<span class="r-in"><span>  features_type <span class="op">=</span> <span class="st">"DE"</span>, feature_source <span class="op">=</span> <span class="st">"both"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:45.728156] Start DEtest</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Workers: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Find all markers(wilcox) among 8 groups...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                                                                      |                                                                                                              |   0%  |                                                                                                                      |==============================================================================================================| 100%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:48.621619] DEtest done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Elapsed time:2.89 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the DE features from AllMarkers_wilcox to calculate distance metric.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DE features number of the query data: 956</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:48.85767] Start DEtest</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Workers: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Find all markers(wilcox) among 13 groups...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                                                                      |                                                                                                              |   0%  |                                                                                                                      |==============================================================================================================| 100%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [2023-11-02 18:14:59.173834] DEtest done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Elapsed time:10.32 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the DE features from AllMarkers_wilcox to calculate distance metric.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DE features number of the ref data: 252</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 67 features to calculate distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected query data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected reference data type: log_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate similarity...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 'raw' method to find neighbors.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Predict cell type...</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> No cell overlap between new meta data and Seurat object</span>
<span class="r-in"><span><span class="fu"><a href="CellDimPlot.html">CellDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, group.by <span class="op">=</span> <span class="st">"KNNPredict_classification"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-12.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="FeatureDimPlot.html">FeatureDimPlot</a></span><span class="op">(</span><span class="va">pancreas_sub</span>, features <span class="op">=</span> <span class="st">"KNNPredict_simil"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunKNNPredict-13.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Hao Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

