<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Single-cell reference mapping with Symphony method"><title>Single-cell reference mapping with Symphony method — RunSymphonyMap • SCP</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Single-cell reference mapping with Symphony method — RunSymphonyMap"><meta property="og:description" content="Single-cell reference mapping with Symphony method"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SCP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/zhanghao-njmu/SCP">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Single-cell reference mapping with Symphony method</h1>
      <small class="dont-index">Source: <a href="https://github.com/zhanghao-njmu/SCP/blob/HEAD/R/SCP-projection.R" class="external-link"><code>R/SCP-projection.R</code></a></small>
      <div class="d-none name"><code>RunSymphonyMap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Single-cell reference mapping with Symphony method</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">RunSymphonyMap</span><span class="op">(</span></span>
<span>  <span class="va">srt_query</span>,</span>
<span>  <span class="va">srt_ref</span>,</span>
<span>  query_assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ref_assay <span class="op">=</span> <span class="va">srt_ref</span><span class="op">[[</span><span class="va">ref_pca</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">assay.used</span>,</span>
<span>  ref_pca <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ref_harmony <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ref_umap <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ref_group <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  projection_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"model"</span>, <span class="st">"knn"</span><span class="op">)</span>,</span>
<span>  nn_method <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  distance_metric <span class="op">=</span> <span class="st">"cosine"</span>,</span>
<span>  vote_fun <span class="op">=</span> <span class="st">"mean"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>srt_query</dt>
<dd><p>An object of class Seurat storing the query cells.</p></dd>


<dt>srt_ref</dt>
<dd><p>An object of class Seurat storing the reference cells.</p></dd>


<dt>query_assay</dt>
<dd><p>A character string specifying the assay name for the query cells. If not provided, the default assay for the query object will be used.</p></dd>


<dt>ref_assay</dt>
<dd><p>A character string specifying the assay name for the reference cells. If not provided, the default assay for the reference object will be used.</p></dd>


<dt>ref_pca</dt>
<dd><p>A character string specifying the name of the PCA reduction in the reference object to use for calculating the distance metric.</p></dd>


<dt>ref_harmony</dt>
<dd><p>A character string specifying the name of the Harmony reduction in the reference object to use for calculating the distance metric.</p></dd>


<dt>ref_umap</dt>
<dd><p>A character string specifying the name of the UMAP reduction in the reference object. If not provided, the first UMAP reduction found in the reference object will be used.</p></dd>


<dt>ref_group</dt>
<dd><p>A character string specifying a metadata column name in the reference object to use for grouping.</p></dd>


<dt>projection_method</dt>
<dd><p>A character string specifying the projection method to use. Options are "model" and "knn". If "model" is selected, the function will try to use a pre-trained UMAP model in the reference object for projection. If "knn" is selected, the function will directly find the nearest neighbors using the distance metric.</p></dd>


<dt>nn_method</dt>
<dd><p>A character string specifying the nearest neighbor search method to use. Options are "raw", "annoy", and "rann". If "raw" is selected, the function will use the brute-force method to find the nearest neighbors. If "annoy" is selected, the function will use the Annoy library for approximate nearest neighbor search. If "rann" is selected, the function will use the RANN library for approximate nearest neighbor search. If not provided, the function will choose the search method based on the size of the query and reference datasets.</p></dd>


<dt>k</dt>
<dd><p>An integer specifying the number of nearest neighbors to find for each cell in the query object.</p></dd>


<dt>distance_metric</dt>
<dd><p>A character string specifying the distance metric to use for calculating the pairwise distances between cells. Options include: "pearson", "spearman", "cosine", "correlation", "jaccard", "ejaccard", "dice", "edice", "hamman", "simple matching", and "faith". Additional distance metrics can also be used, such as "euclidean", "manhattan", "hamming", etc.</p></dd>


<dt>vote_fun</dt>
<dd><p>A character string specifying the function to be used for aggregating the nearest neighbors in the reference object. Options are "mean", "median", "sum", "min", "max", "sd", "var", etc. If not provided, the default is "mean".</p></dd>

</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"panc8_sub"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">srt_ref</span> <span class="op">&lt;-</span> <span class="va">panc8_sub</span><span class="op">[</span>, <span class="va">panc8_sub</span><span class="op">$</span><span class="va">tech</span> <span class="op">!=</span> <span class="st">"fluidigmc1"</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">srt_query</span> <span class="op">&lt;-</span> <span class="va">panc8_sub</span><span class="op">[</span>, <span class="va">panc8_sub</span><span class="op">$</span><span class="va">tech</span> <span class="op">==</span> <span class="st">"fluidigmc1"</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">srt_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="Integration_SCP.html">Integration_SCP</a></span><span class="op">(</span><span class="va">srt_ref</span>, batch <span class="op">=</span> <span class="st">"tech"</span>, integration_method <span class="op">=</span> <span class="st">"Harmony"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-10-30 02:35:19.115332] Start Harmony_integrate</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-10-30 02:35:19.12721] Spliting srtMerge into srtList by column tech... ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-10-30 02:35:19.499202] Checking srtList... ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Data 1/4 of the srtList is raw_normalized_counts. Perform NormalizeData(LogNormalize) on the data ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Perform FindVariableFeatures on the data 1/4 of the srtList...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Data 2/4 of the srtList is raw_normalized_counts. Perform NormalizeData(LogNormalize) on the data ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Perform FindVariableFeatures on the data 2/4 of the srtList...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Data 3/4 of the srtList is raw_counts. Perform NormalizeData(LogNormalize) on the data ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Perform FindVariableFeatures on the data 3/4 of the srtList...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Data 4/4 of the srtList is raw_counts. Perform NormalizeData(LogNormalize) on the data ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Perform FindVariableFeatures on the data 4/4 of the srtList...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Use the separate HVF from srtList...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of available HVF: 2000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-10-30 02:35:23.222902] Finished checking.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-10-30 02:35:23.78261] Perform ScaleData on the data...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-10-30 02:35:23.953184] Perform linear dimension reduction (pca) on the data...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-10-30 02:35:24.635085] Perform integration(Harmony) on the data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Harmony integration using Reduction(Harmonypca, dims:1-10) as input</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Transposing data matrix</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-10-30 02:35:25.775282] Perform FindClusters (louvain) on the data...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-10-30 02:35:25.882061] Reorder clusters...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-10-30 02:35:25.992421] Perform nonlinear dimension reduction (umap) on the data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Non-linear dimensionality reduction(umap) using Reduction(Harmony, dims:1-10) as input</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘spam’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘spam’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Non-linear dimensionality reduction(umap) using Reduction(Harmony, dims:1-10) as input</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘spam’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘spam’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2023-10-30 02:35:42.358017] Harmony_integrate done</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Elapsed time: 23.24 secs </span>
<span class="r-in"><span><span class="fu"><a href="CellDimPlot.html">CellDimPlot</a></span><span class="op">(</span><span class="va">srt_ref</span>, group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"celltype"</span>, <span class="st">"tech"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunSymphonyMap-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Projection</span></span></span>
<span class="r-in"><span><span class="va">srt_query</span> <span class="op">&lt;-</span> <span class="fu">RunSymphonyMap</span><span class="op">(</span>srt_query <span class="op">=</span> <span class="va">srt_query</span>, srt_ref <span class="op">=</span> <span class="va">srt_ref</span>, ref_pca <span class="op">=</span> <span class="st">"Harmonypca"</span>, ref_harmony <span class="op">=</span> <span class="st">"Harmony"</span>, ref_umap <span class="op">=</span> <span class="st">"HarmonyUMAP2D"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected srt_query data type: raw_normalized_counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Detected srt_ref data type: log_normalized_counts</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Data type is unknown or different between srt_query and srt_ref.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Build reference</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saved embeddings</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saved soft cluster assignments</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saved variable gene information for 2000 genes.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saved PCA loadings.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saved metadata.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate final L2 normalized reference centroids (Y_cos)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate reference compression terms (Nr and C)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished nicely.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run mapQuery</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Scaling and synchronizing query gene expression</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 2000 reference variable genes in query dataset</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Project query cells using reference gene loadings</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering query cells to reference centroids</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Correcting query batch effects</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run UMAP projection</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the reduction to calculate distance metric.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 'raw' method to find neighbors.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running UMAP projection</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 02:35:43 Read 200 rows</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 02:35:43 Processing block 1 of 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 02:35:43 Commencing smooth kNN distance calibration using 1 thread</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  with target n_neighbors = 30</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 02:35:43 Initializing by weighted average of neighbor coordinates using 1 thread</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 02:35:43 Commencing optimization for 200 epochs, with 6000 positive edges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 02:35:43 Finished</span>
<span class="r-in"><span><span class="fu"><a href="ProjectionPlot.html">ProjectionPlot</a></span><span class="op">(</span>srt_query <span class="op">=</span> <span class="va">srt_query</span>, srt_ref <span class="op">=</span> <span class="va">srt_ref</span>, query_group <span class="op">=</span> <span class="st">"celltype"</span>, ref_group <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Scale for <span style="color: #00BB00;">x</span> is already present.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding another scale for <span style="color: #00BB00;">x</span>, which will replace the existing scale.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Scale for <span style="color: #00BB00;">y</span> is already present.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding another scale for <span style="color: #00BB00;">y</span>, which will replace the existing scale.</span>
<span class="r-plt img"><img src="RunSymphonyMap-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Hao Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

